### PR TITLE
feat: `roxctl` to connect to central via port-forward

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

Add `--use-current-k8s-context` command line option, which starts a port-forward to the central pod, and sets central endpoint to `https://localhost:<port>`.

The feature should be used by all `roxctl` commands which operate Kubernetes resources AND call central API.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

```console
hack@on:~$ ./bin/linux_amd64/roxctl --help
...
Global Flags:
      --use-current-k8s-context    Use the current kubeconfig context to connect to central via port-forwarding. Alternatively, set ROX_USE_KUBECONTEXT environment variable to true (default false) (default false)
...
hack@on:~$ ./bin/linux_amd64/roxctl --use-current-k8s-context central whoami --insecure-skip-tls-verify
UserID:
	auth-token:faddd24a-878b-4d7e-9598-49d2dd569e9b
...
hack@on:~$ ./bin/linux_amd64/roxctl --use-current-k8s-context central whoami --insecure-skip-tls-verify -p password
UserID:
	admin
User name:
	
Roles:
	- Admin
...
```

### CODE REVIEW

File: roxctl/common/k8s_port_forward.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+
+	forwarder, err := portforward.New(dialer, []string{"0:8443"}, stopChannel, readyChannel, streams.Out, streams.ErrOut)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	go func() {
+		_ = forwarder.ForwardPorts()
+	}()
+	<-readyChannel
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:135]:
```suggestion
  errChan := make(chan error)
	go func() {
		errChan <- forwarder.ForwardPorts()
	}()
	select {
	 case <-readyChannel
	 case err <- errChan:
	   if err != nil {
		   return 0, errPortForwarding.CausedBy(err)
	  }
	}
```

**Comment by connorgorman** [comment]
What is the specific use case we are trying to enable?

**Comment by 0x656b694d** [comment]
> What is the specific use case we are trying to enable?

@connorgorman, avoiding double-headed connection to central:
1. `roxctl` commands that manipulate kubernetes resources must not connect to central via external central endpoint (not sure though we've got any today, but the declarative configuration commands could already benefit from the feature);
2. CLI users who operate their kubernetes clusters may prefer to ensure they're connecting to the same cluster as with `kubectl`;
3. In couple with #7904 enable password-less login.

File: go.mod [hunk 2]
```diff
@@ -206,6 +206,7 @@ require (
 	github.com/facebookgo/stack v0.0.0-20160209184415-751773369052 // indirect
 	github.com/facebookgo/subset v0.0.0-20200203212716-c811ad88dec4 // indirect
 	github.com/facebookincubator/flog v0.0.0-20190930132826-d2511d0ce33c // indirect
+	github.com/fatih/camelcase v1.0.0 // indirect
```
Comment (janisz) [inline go.mod:209]: That's interesting. no other deps were changed and we have new indirect one?

File: roxctl/common/tls.go [hunk 3]
```diff
@@ -56,6 +56,12 @@ func ConnectNames() (string, string, bool, error) {
 	if err != nil {
 		return "", "", false, errors.Wrap(err, "could not get endpoint")
 	}
+	if flags.UseKubeContext() {
```
Comment (janisz) [inline roxctl/common/tls.go:59]: Shouldn't this at the top as this overrides endpoint provided in flag or maybe the two options should be mutually exclusive?

File: roxctl/common/k8s_port_forward.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	stopChannel := make(chan struct{}, 1)
+	readyChannel := make(chan struct{})
+
+	go func() {
+		<-signals
+		if stopChannel != nil {
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:110]: Is it possible to be stopChannel be nil?

File: roxctl/common/k8s_port_forward.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		Resource("pods").
+		Namespace(namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	transport, upgrader, err := spdy.RoundTripperFor(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	streams := genericclioptions.IOStreams{}
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:125]:
```suggestion
	streams := genericiooptions.IOStreams{}
```

File: roxctl/common/k8s_port_forward.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- forwarder.ForwardPorts()
+	}()
+	select {
+	case <-readyChannel:
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:138]: Do we need to wait on `stopChannel` here too?

Comment (janisz) [reply roxctl/common/k8s_port_forward.go:138]: Probably not as stopChannel will cause `ForwardPorts` to return `nil`

File: roxctl/common/k8s_port_forward.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	ports, err := forwarder.GetPorts()
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	return ports[0].Local, nil
+}
+
+// GetForwardingEndpoint starts port-forwarding to svc/central and returns
+// the endpoint with local port forwarding to the service port.
+func GetForwardingEndpoint() (string, error) {
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:154]: unused?

**Review by janisz — CHANGES_REQUESTED** [review]
I think it's not ready yet.
I left some comments. How about adding e2e tests. In theory we should be able to reuse all tests we currently have but parametrize them to use port-forwarder.

File: go.mod (hunk 2 above)
Comment (0x656b694d) [reply go.mod:209]: I guess some new feature is now used from some already existing direct dependency:
```console
$ go mod why github.com/fatih/camelcase
# github.com/fatih/camelcase
github.com/stackrox/rox/roxctl/common
k8s.io/kubectl/pkg/polymorphichelpers
k8s.io/kubectl/pkg/describe
github.com/fatih/camelcase
```

File: roxctl/common/k8s_port_forward.go (hunk 5 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go:125]: Why? What's that `genericiooptions`?

File: roxctl/common/k8s_port_forward.go (hunk 1 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go]: Done

File: roxctl/common/k8s_port_forward.go (hunk 4 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go:110]: Not really.

File: roxctl/common/k8s_port_forward.go (hunk 7 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go:154]: [endpoint, err = GetForwardingEndpoint()](https://github.com/stackrox/stackrox/blob/3e3d93c0af004a3977c01b01b098b28c6d702864/roxctl/common/tls.go#L60)

File: roxctl/common/k8s_port_forward.go (hunk 6 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go:138]: No, we don't need to wait for `stopChannel`. This is to shutdown the port forwarding on a signal.

File: roxctl/common/tls.go (hunk 3 above)
Comment (0x656b694d) [reply roxctl/common/tls.go:59]: It shouldn't. It is here to pick up the `usePlaintext` option.
[c.MarkFlagsMutuallyExclusive(useKubeContextFlagName, "endpoint")](https://github.com/stackrox/stackrox/blob/3e3d93c0af004a3977c01b01b098b28c6d702864/roxctl/common/flags/endpoint.go#L82)

**Review by 0x656b694d — COMMENTED** [review]
Will see how to test e2e.

File: roxctl/common/k8s_port_forward.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	portForwardOnce     sync.Once
+	forwardEndpoint     string
+	portForwardingError error
+
+	errPortForwarding = errox.ServerError.New("port-forwarding error")
+)
+
+// setKubernetesDefaults sets default values on the provided client config for accessing the
+// Kubernetes API or returns an error if any of the defaults are impossible or invalid.
+// TODO this isn't what we want.  Each clientset should be setting defaults as it sees fit.
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:38]: Add issues with description for TODOs

File: roxctl/common/k8s_port_forward.go (hunk 5 above)
Comment (janisz) [reply roxctl/common/k8s_port_forward.go:123]: `genericclioptions.IOStreams{}` is deprecated
https://pkg.go.dev/k8s.io/cli-runtime/pkg/genericclioptions#IOStreams:~:text=and%20review%20code-,DEPRECATED,-%3A%20use%20genericiooptions.IOStreams

**Comment by 0x656b694d** [comment]
> `roxctl` may hang if there is a connection problem after the port is open.

Turned out it has nothing to do with port forward. It retries for 51.2 seconds even after connection refused.

File: roxctl/common/k8s_port_forward.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+	}
+}
+
+func sortByPhase(pods []*corev1.Pod) sort.Interface {
+	return sort.Reverse(podutils.ActivePods(pods))
+}
+
+func getCentralServiceSelectors(coreClient coreclient.ServicesGetter, namespace string) (string, error) {
+	svc, err := coreClient.Services(namespace).
+		Get(context.TODO(), "central", metav1.GetOptions{})
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:67]: why not pass a context?

File: roxctl/common/k8s_port_forward.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	if err != nil {
+		return "", errors.WithMessage(err, "failed to get central service")
+	}
+	selectors := []string{}
+	for k, v := range svc.Spec.Selector {
+		selectors = append(selectors, k+"="+v)
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:73]: do we need to escape or quote k or v?

File: roxctl/common/k8s_port_forward.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	return strings.Join(selectors, ","), nil
+}
+
+//nolint:wrapcheck
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:78]: why?

File: roxctl/common/k8s_port_forward.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	forwardablePod, _, err := polymorphichelpers.GetFirstPod(coreClient,
+		namespace, selectors, 10*time.Second, sortByPhase)
+	if err != nil {
+		return nil, err
+	}
+
+	return coreClient.Pods(namespace).
+		Get(context.TODO(), forwardablePod.GetName(), metav1.GetOptions{})
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:92]: dito

File: roxctl/common/k8s_port_forward.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+//nolint:wrapcheck
+func getCentralPod(coreClient coreclient.CoreV1Interface, namespace string) (*corev1.Pod, error) {
+	selectors, err := getCentralServiceSelectors(coreClient, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	forwardablePod, _, err := polymorphichelpers.GetFirstPod(coreClient,
+		namespace, selectors, 10*time.Second, sortByPhase)
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:86]: w have a timeout flag should we use it? or it's just a safety check?

File: roxctl/common/k8s_port_forward.go [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+		Namespace(pod.GetNamespace()).Name(pod.GetName()).
+		SubResource("portforward")
+
+	transport, upgrader, err := spdy.RoundTripperFor(restConfig)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to configure k8s REST client transport")
+	}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	return portforward.New(dialer, []string{"0:8443"}, stopChannel, readyChannel, nil, nil) //nolint:wrapcheck
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:110]: could you add a comment to a place where it's harcoded

File: roxctl/common/k8s_port_forward.go [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+func getCentralPod(ctx context.Context, coreClient coreclient.CoreV1Interface, namespace string) (*corev1.Pod, error) {
+	selectors, err := getCentralServiceSelectors(ctx, coreClient, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	forwardablePod, _, err := polymorphichelpers.GetFirstPod(coreClient,
+		namespace, selectors, reasonableTimeout, sortByPhase)
+	if err != nil {
+		return nil, err //nolint:wrapcheck
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:88]: I think we could wrap error here
```suggestion
		return nil, errors.Wrap(err, "could not get pod")
```

File: roxctl/common/k8s_port_forward.go [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+	}
+
+	return coreClient.Pods(namespace). //nolint:wrapcheck
+						Get(ctx, forwardablePod.GetName(), metav1.GetOptions{})
+}
+
+func getCentralAPIPort(pod *corev1.Pod) int32 {
+	if pod != nil {
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:96]: too defensive; we discussed, the pod can't be null here we should have a check earlier in the code

File: roxctl/common/k8s_port_forward.go [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, req.URL())
+	port := getCentralAPIPort(pod)
+	return portforward.New(dialer, []string{fmt.Sprintf("0:%d", port)}, stopChannel, readyChannel, nil, nil) //nolint:wrapcheck
+}
+
+func getConfigs() (*rest.Config, coreclient.CoreV1Interface, string, error) {
+	kubeConfigLoader := genericclioptions.NewConfigFlags(true).ToRawKubeConfigLoader()
+	namespace, _, err := kubeConfigLoader.Namespace()
+	if err != nil {
+		return nil, nil, "", errors.WithMessage(err, "failed to identify central namespace")
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:132]: Why not wrapping?

File: roxctl/common/k8s_port_forward.go [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+	select {
+	case <-readyChannel:
+	case err := <-errChan:
+		if err != nil {
+			return 0, err
+		}
+	}
+	ports, err := forwarder.GetPorts()
+	if err != nil {
+		return 0, errors.WithMessage(err, "failed to aquire forwarding ports")
```
Comment (janisz) [inline roxctl/common/k8s_port_forward.go:198]:
```suggestion
		return 0, errors.WithMessage(err, "failed to acquire forwarding ports")
```

**Review by janisz — APPROVED** [review]
just nitpics?

File: roxctl/common/k8s_port_forward.go (hunk 17 above)
Comment (0x656b694d) [reply roxctl/common/k8s_port_forward.go:132]: Because we don't need stack.

**Pull request merged**